diff --git a/src/drivers/driver_athr.c b/src/drivers/driver_athr.c
index 2cc0bf8..30f5e1c 100644
--- a/src/drivers/driver_athr.c
+++ b/src/drivers/driver_athr.c
@@ -2341,7 +2341,7 @@ driver_atheros_del_key(struct driver_atheros_data *drv, int key_idx,
 
 int
 driver_atheros_set_key(const char *ifname, void *priv, enum wpa_alg alg,
-		       const u8 *addr, int key_idx, int set_tx,
+		       const u8 *addr, int key_idx, int vlan_id, u8 vlan_found, int set_tx,
 		       const u8 *seq, size_t seq_len,
 		       const u8 *key, size_t key_len)
 {
